source/common/grpc/async_client_impl.cc

```diff
@@ -95,8 +95,16 @@ void AsyncStreamImpl::initialize(bool buffer_body_for_retry) {
       parent_.host_name_.empty() ? parent_.remote_cluster_name_ : parent_.host_name_,
       service_full_name_, method_name_, options_.timeout);
   // Fill service-wide initial metadata.
-  parent_.metadata_parser_->evaluateHeaders(headers_message_->headers(),
-                                            options_.parent_context.stream_info);
+  // TODO(cpakulski): Find a better way to access requestHeaders after runtime guard
+  // envoy_reloadable_features_unified_header_formatter runtime guard is deprecated.
+  // Maybe put it to parent_context?
+  parent_.metadata_parser_->evaluateHeaders(
+      headers_message_->headers(),
+      ((options_.parent_context.stream_info != nullptr) &&
+       (options_.parent_context.stream_info->getRequestHeaders() != nullptr))
+          ? *options_.parent_context.stream_info->getRequestHeaders()
+          : *Http::StaticEmptyHeaders::get().request_headers,
```

Comment:
If it's possible for request headers to be empty, why not just make the argument an OptRef?

Reply:
Agree. As I noted in TODO section just above those lines, that section needs to be reworked after runtime guard is deprecated. I am going to add the suggestion to use optref to that TODO comment.